File: aoc.net/Aoc.Tests/2023/01/Code.cs
// Copyright © 2023 TradingLens. All Rights Reserved.
using Xunit;
using Xunit.Abstractions;

namespace Aoc.Tests._2023._01; 

public class Code {
  public ITestOutputHelper Output { get; }
  public Code(ITestOutputHelper output) {
    Output = output;
  }
  private int? GetDigit(int i, string str, bool allowWordDigits) {
    if (char.IsDigit(str[i])) return str[i] - '0';
    if (allowWordDigits) {
      if (i < 2) return null;
      switch (str.Substring(i-2,3)) {
          case "one": return 1;  
          case "two": return 2;  
          case "six": return 6;  
      }
      if (i >= 3) { 
        switch (str.Substring(i-3,4)) {
          case "four": return 4;  
          case "five": return 5;  
          case "nine": return 9;  
        }
      }
      if (i >= 4) { 
        switch (str.Substring(i-4,5)) {
          case "three": return 3;  
          case "seven": return 7;  
          case "eight": return 8;  
        }
      }
    }
    return null;
  }

  public int ParseLine(string str, bool allowWordDigits)
  {
    var val = 0;
    for (int i = 0; i < str.Length; i++) {
      var digit = GetDigit(i, str, allowWordDigits);
      if (digit != null) {
        val += 10 * digit.Value;
        break;
      }
    }
    for (int i = str.Length - 1; i >= 0; i--)
    {
      var digit = GetDigit(i, str, allowWordDigits);
      if (digit != null) {
        val += digit.Value;
        break;
      }
    }
    return val;
  }  
  [Fact]
  public void Run() {
    Output.WriteLine(File.ReadLines("2023/01/example2.txt").Sum(x => ParseLine(x,true)).ToString());
    Output.WriteLine(File.ReadLines("2023/01/input.txt").Sum(x => ParseLine(x,false)).ToString());
    Output.WriteLine(File.ReadLines("2023/01/input.txt").Sum(x => ParseLine(x,true)).ToString());
  }
}

File: python/2023/01/code.py
import os
sd = os.path.dirname(os.path.abspath(__file__))

def getworddigit(l, i):
  if(i >= 2):
    if(l[i-2:i+1] == "one"):  return 1
    if(l[i-2:i+1] == "two"):  return 2 
    if(l[i-2:i+1] == "six"):  return 6
  if(i >= 3):
    if(l[i-3:i+1] == "four"):  return 4
    if(l[i-3:i+1] == "five"):  return 5 
    if(l[i-3:i+1] == "nine"):  return 9
  if(i >= 4):
    if(l[i-4:i+1] == "three"):  return 3
    if(l[i-4:i+1] == "seven"):  return 7 
    if(l[i-4:i+1] == "eight"):  return 8
  return -1

def addLine(l, allow_word_digits = False):
  result = 0
  length = len(l)
  for i in range(0, length):
    ch = l[i]
    if(ch.isdigit()): 
       result += 10*(ord(ch)-ord('0'))
       break
    if(allow_word_digits):
      val = getworddigit(l, i)
      if(val >= 0):
        result += 10*val
        break
  for i in range(length - 1, -1, -1):
    ch = l[i]
    if(ch.isdigit()): 
       result += ord(ch)-ord('0')
       break
    if(allow_word_digits):
      val = getworddigit(l, i)
      if(val >= 0):
        result += val
        break
  return result

def evaluatefile(filepath, allow_word_digits = False):
  result = 0
  with open(filepath) as file:
    for line in file:
      result += addLine(line, allow_word_digits)
  return result
    
print(evaluatefile(sd+"/example1.txt"))
print(evaluatefile(sd+"/input1.txt"))
print(evaluatefile(sd+"/example2.txt", False))
print(evaluatefile(sd+"/example2.txt", True))
print(evaluatefile(sd+"/input1.txt", True))